File: packaging/mcpb/kotlin/stackql-mcp/src/main/kotlin/io/stackql/mcp/Pins.kt
package io.stackql.mcp

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * The pinned stackql release, download base URL and per-platform sha256 pins.
 *
 * Nothing is hand-written here: the values come from the `platforms.json`
 * classpath resource, rendered by `packaging/mcpb/scripts/render-platforms.sh`
 * from the published `.mcpb.sha256` release assets - the same manifest every
 * other StackQL wrapper (npm, PyPI, the other SDKs) ships. Render it before
 * building: `make kotlin-manifest VERSION=X.Y.Z` (from packaging/mcpb).
 */
object Pins {
    private const val RESOURCE = "/platforms.json"

    @Serializable
    private data class Manifest(
        val version: String,
        val baseUrl: String,
        val platforms: Map<String, PlatformPin>,
    )

    @Serializable
    data class PlatformPin(val bundle: String, val sha256: String)

    private val manifest: Manifest by lazy {
        val stream = Pins::class.java.getResourceAsStream(RESOURCE)
            ?: throw StackqlMcpException(
                "classpath resource $RESOURCE not found - render it with " +
                    "'make kotlin-manifest VERSION=X.Y.Z' from packaging/mcpb",
            )
        val text = stream.use { it.readBytes().toString(Charsets.UTF_8) }
        Json { ignoreUnknownKeys = true }.decodeFromString<Manifest>(text).also {
            require(it.version.isNotBlank() && it.baseUrl.isNotBlank() && it.platforms.isNotEmpty()) {
                "platforms.json is missing version, baseUrl or platforms"
            }
        }
    }

    /** The stackql release this library is version-locked to (leading `v` stripped). */
    val STACKQL_VERSION: String get() = manifest.version

    /** Front door bundles are downloaded from (`https://releases.stackql.io/stackql/<version>`). */
    val BASE_URL: String get() = manifest.baseUrl

    /** User-Agent sent to the download proxy, per vector and version. */
    val USER_AGENT: String get() = "stackql-mcp-server-kotlin/${manifest.version}"

    /** Lowercase hex sha256 of each platform's `.mcpb` bundle for [STACKQL_VERSION]. */
    val pins: Map<Platform, String> by lazy {
        Platform.entries.associateWith { platform ->
            manifest.platforms[platform.key]?.sha256
                ?: throw StackqlMcpException("platforms.json has no pin for ${platform.key}")
        }
    }

    /** The published pin for [platform]; every platform has one. */
    fun pinFor(platform: Platform): String =
        pins[platform] ?: throw StackqlMcpException("no pin for platform ${platform.key}")

    /** Download URL for [platform]'s pinned `.mcpb` bundle: `<baseUrl>/<bundle>`. */
    fun bundleUrl(platform: Platform): String {
        val bundle = manifest.platforms[platform.key]?.bundle ?: platform.bundleName
        return "$BASE_URL/$bundle"
    }
}

File: packaging/mcpb/swift/Sources/StackQLMCP/Pins.swift
import Foundation

/// Version, download base URL and sha256 pins for the StackQL MCP server
/// bundle.
///
/// Nothing is hand-written here: the values are the `platforms.json` package
/// resource, rendered by `packaging/mcpb/scripts/render-platforms.sh` from the
/// published `.mcpb.sha256` release assets - the same manifest every other
/// StackQL wrapper (npm, PyPI, the other SDKs) ships. Render it before
/// building: `make swift-manifest VERSION=X.Y.Z` (from packaging/mcpb).
public enum Pins {
    /// One platform's bundle name and lowercase-hex sha256.
    public struct PlatformPin: Sendable, Decodable {
        public let bundle: String
        public let sha256: String
    }

    /// The parsed platforms.json.
    public struct Manifest: Sendable, Decodable {
        /// The stackql release this package is version-locked to (no leading v).
        public let version: String
        /// Front door bundles are downloaded from
        /// (`https://releases.stackql.io/stackql/<version>`).
        public let baseUrl: String
        /// Platform key -> pin.
        public let platforms: [String: PlatformPin]
    }

    /// The embedded manifest. Missing or malformed is a build/packaging error,
    /// not a runtime condition, so it is fatal.
    public static let manifest: Manifest = {
        guard let url = Bundle.module.url(forResource: "platforms", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let manifest = try? JSONDecoder().decode(Manifest.self, from: data),
              !manifest.version.isEmpty, !manifest.baseUrl.isEmpty, !manifest.platforms.isEmpty
        else {
            fatalError(
                "stackql mcp: platforms.json resource missing or invalid - render it with "
                    + "'make swift-manifest VERSION=X.Y.Z' from packaging/mcpb")
        }
        return manifest
    }()

    /// The stackql release this package is version-locked to.
    public static var defaultVersion: String { manifest.version }

    /// Front door bundles are downloaded from.
    public static var baseURL: String { manifest.baseUrl }

    /// User-Agent sent to the download proxy, per vector and version.
    public static var userAgent: String { "stackql-mcp-server-swift/\(manifest.version)" }

    /// sha256 of the release `.mcpb` bundle for `defaultVersion`, keyed by
    /// platform. Lowercase hex.
    public static var bundleSHA256: [Platform: String] {
        var out: [Platform: String] = [:]
        for platform in Platform.allCases {
            if let pin = manifest.platforms[platform.rawValue] {
                out[platform] = pin.sha256.lowercased()
            }
        }
        return out
    }

    /// The pin for a platform, or nil when none is published.
    public static func pin(for platform: Platform) -> PlatformPin? {
        manifest.platforms[platform.rawValue]
    }

    /// The bundle asset name for a platform (`stackql-mcp-<key>.mcpb`).
    public static func bundleName(_ platform: Platform) -> String {
        pin(for: platform)?.bundle ?? "stackql-mcp-\(platform.rawValue).mcpb"
    }

    /// The download URL for a platform's bundle: `<baseUrl>/<bundle>`.
    public static func bundleURL(_ platform: Platform) -> URL {
        URL(string: "\(baseURL)/\(bundleName(platform))")!
    }
}
